Stock watchlist improvements

kamalq97 opened the pull request:
Changelog:
- Combined `watchlist_add` and `watchlist_remove` into a new url/view called `watchlist_edit` in **urls.py** and **views.py**.
- New method for error handling `watchlist_edit_view` when stock symbol is not found (similar to how errors are handled in the `single_stock` view)
- Added `getCookie` function to **base.html** (mainly used for fetching the "csrftoken" cookie needed for post requests)
- Added JavaScript code to **index.html**, **single_stock.htm**l, and **watchlist.html**. The code executes on the click of the blue or red watchlist button.

 _Now, adding/removing a stock from the watchlist no longer forces the entire webpage to reload._

kamalq97 committed:
Update urls.py, views.py, and html templates

kamalq97 committed:
Merge branch 'master' into StockProject-userprofile

murradkh commented on myapp/templates/index.html:
```diff
@@ -33,21 +33,38 @@
+
+{% block scripts %}
+<script>
+    function watchlistEdit(symbol, number) {
```
Nice Kamal, you learned AJAX very quickly!

kamalq97 replied:
Thanks!

murradkh commented on myapp/templates/index.html:
```diff
@@ -33,21 +33,38 @@
+            icon.className = "fa fa-eye";
+            editButton.className = "btn btn-outline-primary";     
+        }
+        xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));
```
i think you don't need for `getCookie` method to fetch the cookie, you can use `xhr.withCredentials = true;` to include the cookie into the post request
reference: [URL1](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials), [URL2](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS)


let me know if you stuck with it

kamaql97 replied:
I will give it a try

kamalq97 replied:
Unfortunately, the method you talk about did not work, it gave the following error message: 
`Forbidden (CSRF token missing or incorrect.)`

murradkh replied:
i will take a look

murradkh replied:
looks like there is no shortage/alternative way other than what you did to render the `csrfmiddlewaretoken` value
good to go with this.

murradkh commented on myapp/templates/index.html:
```diff
@@ -33,21 +33,38 @@
+            editButton.className = "btn btn-outline-primary";     
+        }
+        xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));
+        xhr.send();
```
what would happen if the request failed for some reason? is the icon would be changed anyway? 
preferred to have a callback function, to be called once the response received from the server.

kamalq97 replied:
Good point!

kamalq97 replied:
@murradkh Fixed! Check my newest commit

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -124,6 +118,23 @@
+            editButton.innerHTML = "<i class='fa fa-eye'></i> Add to watchlist";
+            editButton.className = "btn btn-outline-primary";            
+        }
+        xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));
```
the same

kamalq97 replied:
@murradkh Fixed! Check my newest commit

murradkh replied:
okay

murradkh commented on myapp/urls.py:
```diff
@@ -10,8 +10,7 @@
 	path('accounts/watchlist/', views.watchlist_view, name='watchlist'),
-	path('stock/<str:symbol>/wremove/', views.watchlist_remove_view, name='watchlist_remove'),
-	path('stock/<str:symbol>/wadd/', views.watchlist_add_view, name='watchlist_add'),
+	path('stock/<str:symbol>/<str:operation>/', views.watchlist_edit_view, name='watchlist_edit'),
```
Nice!

murradkh commented on myapp/views.py:
```diff
@@ -88,28 +88,20 @@ def watchlist_view(request):
+        if operation == "wadd":
+            Stock.add_to_watchlist(profile, symbol)
+        nextpage = request.POST.get('next', '/')
+        return redirect(nextpage)               # NOTE: last two lines no longer needed, remove?
```
I'm not sure why "next" is needed and where is assigned to the value

kamaql97 replied:
It is no longer needed, lines 103 and 104 can be safely deleted

kamalq97 replied:
It is no longer needed, we can return some other response

murradkh replied:
you can change to `return HttpResponse("OK")`  (redirecting cause to create redundant request)

murradkh reviewed (commented):
Good Job, many things covered

kamalq97 committed:
Add xhr callback function

murradkh reviewed (approved):
looks better! 
good to merge after you resolve the conflicts

kamalq97 commented:
Yes, I will resolve the conflicts 👍

kamalq97 committed:
Update views.py

Modified retured response

kamalq97 committed:
Merge branch 'master' into StockProject-userprofile

kamalq97 committed:
Update index.html

Added padding around "Stocks per page" button to ensure it is vertically centered in the div

kamalq97 committed:
Update views.py

kamalq97 committed:
Merge branch 'master' into StockProject-userprofile

kamalq97 committed:
Update views.py

Code now passes in all tests

kamalq97 committed:
Update urls.py and views.py
